### PR TITLE
Add reboot service

### PR DESCRIPTION
**Pull request by DurgNomis-drol**
This adds a reboot service. 😄 

It is also possible to do a factory reset from the same endpoint but this is not implemented. I can change this if we want this?

I have tested it locally and it works for me 🚀

### CODE REVIEW

File: custom_components/google_home/sensor.py [hunk 1]
```diff
@@ -131,6 +140,16 @@ def get_device_attributes(device: GoogleHomeDevice) -> DeviceAttributes:
             "integration": DOMAIN,
         }
 
+    async def async_reboot_device(self) -> None:
+        """Reboot device"""
```
Comment (KapJI) [inline custom_components/google_home/sensor.py:144]:
```suggestion
        """Reboot the device."""
```

File: custom_components/google_home/api.py [hunk 2]
```diff
@@ -313,7 +314,66 @@ async def delete_alarm_or_timer(
...
+                        response.status,
+                        response,
+                    )
+        except ClientConnectorError:
+            _LOGGER.warning(
+                "Failed to connect to %s device. The device is probably offline.",
+                device.name,
+            )
+        except ClientError as ex:
+            # Make sure that we log the exception from the client if one occurred.
```
Comment (KapJI) [inline custom_components/google_home/api.py:376]: These services have a lot of duplicated code. Can it be abstracted with a common function?

**Review by Drakulix — APPROVED** [review]
Our api.py really gets a bit repetitive, but that is out of scope for this feature.

Logic seems correct and checks are passing, looks fine to me!

File: custom_components/google_home/api.py [hunk 3]
```diff
@@ -313,7 +314,66 @@ async def delete_alarm_or_timer(
...
         except ClientError as ex:
-            # Make sure that we log the exception if one occurred.
+            # Make sure that we log the exception from the client if one occurred.
+            _LOGGER.error(
+                "Request error: %s",
+                ex,
+            )
+
+    async def reboot_google_device(self, device: GoogleHomeDevice) -> None:
+        """Reboots a Google Home device if supported."""
```
Comment (KapJI) [inline custom_components/google_home/api.py:324]:
```suggestion
        """Reboots a Google Home device if it supports this."""
```

File: custom_components/google_home/api.py [hunk 4]
```diff
@@ -313,7 +314,66 @@ async def delete_alarm_or_timer(
...
+
+        if device.ip_address is None:
+            _LOGGER.warning("Device %s doesn't have IP address!", device.name)
+            return
+
+        if device.auth_token is None:
+            _LOGGER.warning("Device %s doesn't have a auth token!", device.name)
```
Comment (KapJI) [inline custom_components/google_home/api.py:331]:
```suggestion
            _LOGGER.warning("Device %s doesn't have an auth token!", device.name)
```

**Review by KapJI — CHANGES_REQUESTED** [review]

File: custom_components/google_home/api.py (hunk 2 above)
Comment (DurgNomis-drol) [reply custom_components/google_home/api.py:376]: Yes, i thought that too, i even have another thing planned. Should i refactor it in this or should we move it to another PR?

Comment (KapJI) [reply custom_components/google_home/api.py:376]: I'd rather have it here as this adds a new copy.

**Comment by KapJI** [comment]
If you decide to go with factory reset service, I think it should have a confirmation field, something like `confirmation = "i_want_to_reset"`.

**Comment by DurgNomis-drol** [comment]
@KapJI The reason I haven't added it is exactly because of that scenario. I don't think that it is necessary to add and only mentioned it if some of you had another opinion 😄 But yes, that would be the correct way to do it if so.

**Comment by KapJI** [comment]
I think we can wait before we see someone who needs this for some reason 🙂

File: custom_components/google_home/api.py (hunk 2 above)
Comment (Drakulix) [reply custom_components/google_home/api.py:376]: Not sure, if it is really worth it. I want to start replacing api.py with googledevices, once I have finished up another project. While I agree, that some amount de-duplication would likely be beneficial, we should not invest too much here.

Comment (KapJI) [reply custom_components/google_home/api.py:376]: Well, I think in any case we should try to keep the code here good. I don't think this deduplication should be very hard.

Comment (DurgNomis-drol) [reply custom_components/google_home/api.py:376]: It is fine, i am nearly done with it. We/you can also use this when making `googledevices`. So I don't see it as a total waste of time. 😄 I am also gonna add a "Do not disturb switch" soon.

**Comment by DurgNomis-drol** [comment]
It is giving me a real headache that we are using strict typing. We can't use `any`, so it is very difficult to make generic function that can handle post requests and return a response that we don't know what contains.

**Comment by KapJI** [comment]
You can enable `Any` for a single line/function if you need it. I can figure out later how to get rid of it.

**Comment by DurgNomis-drol** [comment]
@KapJI how do you do that?

**Comment by KapJI** [comment]
Add `# type: ignore[misc]` on that line.

**Comment by DurgNomis-drol** [comment]
I am having some problems with how to handle the response message from the server. I think i am close to figuring it out, but will push it tomorrow 😪

**Comment by KapJI** [comment]
You can push unfinished version, maybe I can help.

**Comment by DurgNomis-drol** [comment]
I actually just got it working, had to check mime_type. But it should work now. 😄 

It is logging `NotImplementedError` for `remove_service` every time you call `reboot_device`, is there a fix in the works over at `glocaltokens`?

I removed `charset=UTF8` from headers as i don't think we need it, just a heads up.

**Comment by KapJI** [comment]
`remove_service` was added in https://github.com/leikoilja/glocaltokens/pull/53, we just need to update the dependency.

@leikoilja let's merge https://github.com/leikoilja/glocaltokens/pull/63 and publish the new version.

File: custom_components/google_home/api.py [hunk 5]
```diff
@@ -236,73 +237,108 @@ async def delete_alarm_or_timer(
...
+        if response:
+            # It will return true even if the device does not support rebooting.
+            _LOGGER.info(
+                "Successfully asked %s to reboot.",
+                device.name,
+            )
+
+    async def post(  # type: ignore[misc]
+        self, endpoint: str, data: str, device: GoogleHomeDevice
+    ) -> Any:
```
Comment (KapJI) [inline custom_components/google_home/api.py:306]: I think this should work. Should be enough precision and better that `Any`.
```suggestion
    ) -> Optional[Dict[str, str]]:
```

File: custom_components/google_home/const.py [hunk 6]
```diff
@@ -54,13 +55,14 @@
...
 API_ENDPOINT_DELETE = "setup/assistant/alarms/delete"
+API_ENDPOINT_REBOOT = "setup/reboot"
 HEADER_CAST_LOCAL_AUTH = "cast-local-authorization-token"
 HEADER_CONTENT_TYPE = "content-type"
 
 # HEADERS
 HEADERS = {
     HEADER_CAST_LOCAL_AUTH: "",
-    HEADER_CONTENT_TYPE: "application/json; charset=UTF-8",
+    HEADER_CONTENT_TYPE: "application/json",
```
Comment (KapJI) [inline custom_components/google_home/const.py:65]: Probably this is default but I remember you said you can assign name to an alarm? Can you check if UTF-8 name works fine without setting `charset`?

File: custom_components/google_home/api.py (hunk 5 above)
Comment (DurgNomis-drol) [reply custom_components/google_home/api.py:306]: I will test it 😊

Comment (DurgNomis-drol) [reply custom_components/google_home/api.py:306]: Worked! Thanks for the help

File: custom_components/google_home/const.py (hunk 6 above)
Comment (DurgNomis-drol) [reply custom_components/google_home/const.py:65]: It works, i even tried with the special letters we have here in Denmark and it works just fine. It was maybe me who added it sometime ago with out fully understanding that it was not needed. 😄 

[`Ghlocalapi`](https://rithvikvibhu.github.io/GHLocalApi/#section/Google-Home-Local-API/Getting-Started) also does not say anything about it 😄

File: custom_components/google_home/api.py [hunk 7]
```diff
@@ -236,73 +237,108 @@ async def delete_alarm_or_timer(
...
             url,
-            data,
         )
 
+        resp = None
+
         try:
             async with self._session.post(
-                url, json=data, headers=HEADERS, timeout=TIMEOUT
+                url, data=data, headers=HEADERS, timeout=TIMEOUT
```
Comment (KapJI) [inline custom_components/google_home/api.py:332]: Why did you change `json` to `data`? Is it also works?

Comment (KapJI) [reply custom_components/google_home/api.py:332]: Documentation says that it only adds "content-type: application/json", but I think setting headers explicitly redefines this anyway so it should be fine.

**Review by KapJI — APPROVED** [review]
Looks great, thanks! 🎉

File: custom_components/google_home/api.py (hunk 7 above)
Comment (DurgNomis-drol) [reply custom_components/google_home/api.py:332]: `json` also converts it to JSON as far as i understand.

It is because of strict typing. If we use `json` we need to pass a known dict and we don't know this. This way we only need to define it to a `str`.

But maybe i am wrong about this. 🤔

**Comment by leikoilja** [comment]
Awesome job guys, @DurgNomis-drol, @KapJI! Thanks for implementing it 💥 
I'll make a new release of `glocaltokens` today and open a PR here to bump it up as well ;)

**Comment by DurgNomis-drol** [comment]
@leikoilja Awesome! 💥

File: custom_components/google_home/api.py (hunk 7 above)
Comment (KapJI) [reply custom_components/google_home/api.py:332]: I see. Yeah, that's not easy to construct proper type for json and mypy still doesn't support recursive types which are needed for the correct type.

Comment (DurgNomis-drol) [reply custom_components/google_home/api.py:332]: This is a common problem i am running into with different libraries unfortunately. 😢 But by finding a fix I only get better at understanding python. 🚀 Also many places they use `Any` instead of explicitly stating the correct type.
